baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreferCollectionConstructors.java

```diff
@@ -0,0 +1,345 @@
+                    .put(NEW_HASH_SET, HashSet.class)
+                    .put(NEW_HASH_SET_WITH_ITERABLE, HashSet.class)
+                    .put(NEW_HASH_SET_WITH_EXPECTED_SIZE, HashSet.class)
+                    .put(NEW_TREE_MAP, TreeMap.class)
+                    .put(NEW_TREE_MAP_WITH_SORTED_MAP, TreeMap.class)
+                    .put(NEW_TREE_MAP_WITH_COMPARATOR, TreeMap.class)
+                    .put(NEW_LINKED_HASH_MAP, LinkedHashMap.class)
+                    .put(NEW_LINKED_HASH_MAP_WITH_MAP, LinkedHashMap.class)
+                    .put(NEW_LINKED_HASH_MAP_WITH_EXPECTED_SIZE, LinkedHashMap.class)
+                    .put(NEW_ENUM_MAP, EnumMap.class)
+                    .put(NEW_ENUM_MAP_WITH_CLASS, EnumMap.class)
+                    .put(NEW_ENUM_MAP_WITH_MAP, EnumMap.class)
+                    .put(NEW_IDENTITY_HASH_MAP, IdentityHashMap.class)
+                    .build();
+
+    private static final Set<Matcher<ExpressionTree>> requiresCollectionArg = ImmutableSet.of(
+            NEW_ARRAY_LIST_WITH_ITERABLE,
+            NEW_LINKED_LIST_WITH_ITERABLE,
+            NEW_COPY_ON_WRITE_ARRAY_LIST_WITH_ITERABLE,
+            NEW_COPY_ON_WRITE_ARRAY_SET_WITH_ITERABLE,
+            NEW_LINKED_HASH_SET_WITH_ITERABLE,
+            NEW_TREE_SET_WITH_ITERABLE,
+            NEW_HASH_SET_WITH_ITERABLE);
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+
+        Class<?> collectionClass = findCollectionClassToUse(state, tree);
+        if (collectionClass == null) {
+            return Description.NO_MATCH;
+        }
+
+        SuggestedFix.Builder fixBuilder = SuggestedFix.builder();
+        String collectionType = SuggestedFixes.qualifyType(state, fixBuilder, collectionClass.getName());
+        String typeArgs = tree.getTypeArguments()
+                .stream()
+                .map(state::getSourceForNode)
+                .collect(Collectors.joining(", "));
+        String arg = tree.getArguments().isEmpty() ? "" : state.getSourceForNode(tree.getArguments().get(0));
+        String replacement = "new " + collectionType + "<" + typeArgs + ">(" + arg + ")";
+        return buildDescription(tree)
+                .setMessage("The factory method call should be replaced with a constructor call. "
+                        + "See https://git.io/JeCT6 for more information.")
+                .addFix(fixBuilder.replace(tree, replacement).build())
+                .build();
+    }
+
+    private Class<?> findCollectionClassToUse(VisitorState state, ExpressionTree tree) {
+        boolean firstArgIsCollection = isFirstArgCollection(state, tree);
+        for (Map.Entry<Matcher<ExpressionTree>, Class<?>> entry : classMap.entrySet()) {
+            Matcher<ExpressionTree> matcher = entry.getKey();
+            if (matcher.matches(tree, state) && (!requiresCollectionArg.contains(matcher) || firstArgIsCollection)) {
```

Comment:
If matcher.matches is true and the second check is false, I think null should be returned. Are there any cases where we would want to check more cases?

Reply:
Right. As of now, all matchers happen to be mutually exclusive, so that should be a safe optimization. I'll update.